pulsar-client-api/src/main/java/org/apache/pulsar/client/internal/ReflectionUtils.java

```diff
@@ -50,7 +50,15 @@
     @SuppressWarnings("unchecked")
     static <T> Class<T> newClassInstance(String className) {
         try {
-            return (Class<T>) DefaultImplementation.class.getClassLoader().loadClass(className);
+            try {
+                // when the API is loaded in the same classloader as the impl
+                return (Class<T>) DefaultImplementation.class.getClassLoader().loadClass(className);
```

Comment:
Can't we just use `Thread.currentThread().getContextClassLoader().loadClass(className);` ?

Reply:
So there can be scenario in which pulsar-client-api and pulsar-client-impl is loaded in a classloader that is NOT the context classloader for whatever reason.  For BC purposes lets first check if pulsar-client-impl is loaded in the same classloader as the api and then we can check the context classloader.

This problem was discovered because one of tests in the pulsar-kafka-compat module failed due do powermock loading pulsar-client-api and impl in the powermock classloader and not the context/application classloader.  

In theory a user can do that as well.